drivers/soundwire/sysfs_slave_dpn.c

```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2015-2020 Intel Corporation.
+
+#include <linux/device.h>
+#include <linux/mod_devicetable.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include "bus.h"
+#include "sysfs_local.h"
+
+/*
+ * dpN tree sysfs
+ *
+ */
+
+struct sdw_dpn_tree {
+	struct kobject **nodes;
+};
+
+struct dpn_attribute;
+
+struct dpn_attribute {
+	struct attribute	attr;
+	ssize_t (*show)(struct sdw_slave *slave, int N, int dir,
+			struct dpn_attribute *attr, char *buf);
+};
+
+static int get_port_N_dir(struct kobject *kobj, int *N, int *dir,
+			  struct sdw_slave **slv)
+{
+	struct device *dev = kobj_to_dev(kobj->parent);
+	char *nodename;
+	char *rest;
+	char *token;
+	ssize_t ret;
+
+	/* the node name is dp<N>_<sink/src> */
+	nodename = kstrdup(kobj->name, GFP_KERNEL);
+	if (!nodename)
+		return -ENOMEM;
+
+	rest = nodename;
+	token = nodename;
+
+	/* skip dp */
+	token = strsep(&rest, "p");
+	if (!token) {
```

Comment:
I've never used `strsep()` before so I went and read the docs :-) This doesn't look like a correct check to me. `strsep()` only returns `NULL` is the first argument points to `NULL` (i.e. `rest == NULL` in this case) which is literally impossible here. Same holds below of course. I'd go with the good ol' `strchr()` since we're only searching for one character in each call. Actually, if we're sure, that the string is of the above form, I'd be very dull here and do something like `rest += strlen("dp");` And only then replace the "_".

Reply:
I hate working with strings. This sequence works. If you want a change, please provide a diff.

Reply:
@plbossart let's try
```
char *underscore;
if (strncmp(rest, "dp", strlen("dp"))) {
	ret = -EINVAL;
	goto err;
}
token = rest + strlen("dp");
/* extract N */
underscore = strchr(token, '_');
if (!underscore) {
	ret = -EINVAL;
	goto err;
}
*underscore = '\0';
```
obviously neither runtime- nor even compile-tested. If we wanted to be paranoid we could also check that the complete substring between "dp" and '_' only consists of digits but in fact it looks like `kstrtoint()` would actually return an error if non-digits are encountered.

Reply:
no longer needed